File: InterviewBit/TwoPointer/3Sum0/3sum0.cpp
vector<vector<int> > Solution::threeSum(vector<int> &A) {
    vector<vector<int>> res;
    vector<int> triplet;
    vector<int> temp;
    map<vector<int>, bool> present;
    sort(A.begin(),A.end());
    int left, right, n=A.size(), sum;
    for (int i=0;i<n-2;i++){
        if (A[i]>0) break;
        left = i+1;
        right = n-1;
        while(left<right){
            sum = A[i] + A[left]+A[right];
            if (sum==0){
                triplet.push_back(A[i]);
                triplet.push_back(A[left]);
                triplet.push_back(A[right]);
                if (present.find(triplet)==present.end()) {
                    res.push_back(triplet);
                    present[triplet]=true;
                }
                triplet.clear();
            }
            if (sum>0) right --;
            else left++;
        }
    }
    return res;
}

File: InterviewBit/TwoPointer/3Sum0/3sum0.py
from collections import defaultdict
class Solution:
    # @param A : list of integers
    # @return a list of list of integers
    def threeSum(self, A):
        res = [0]
        n = len(A)
        find = defaultdict(lambda:False)
        A.sort()
        for i in range(n-2):
            left = i+1
            right = n-1
            if A[i]>0: 
                break
            while(left<right):
                triplet = (A[i],A[left],A[right])
                sum_a = sum(triplet)
                if (sum_a==0):
                    if not find[triplet]:
                        find[triplet]=True
                        res.append(triplet)
                if sum_a<0:
                    left+=1
                else:
                    right-=1
        return res[1:]